src/v/raft/append_entries_buffer.cc

```diff
@@ -0,0 +1,118 @@
+#include "raft/append_entries_buffer.h"
+
+#include "raft/consensus.h"
+#include "raft/types.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/util/later.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <exception>
+#include <variant>
+#include <vector>
+namespace raft {
+
+append_entries_buffer::append_entries_buffer(
+  consensus& c, size_t max_buffered_elements)
+  : _consensus(c)
+  , _max_buffered(max_buffered_elements) {}
+
+ss::future<append_entries_reply>
+append_entries_buffer::enqueue(append_entries_request&& r) {
+    return ss::with_gate(_gate, [this, r = std::move(r)]() mutable {
+        // we normally do not want to wait as it would cause requests
+        // reordering, we use non blocking API of the semaphore.
+        ss::promise<append_entries_reply> p;
+        auto f = p.get_future();
+
+        if (_requests.size() >= _max_buffered) {
+            return ss::make_exception_future<append_entries_reply>(
+              std::overflow_error("max buffered entries count reached."));
+        }
+
+        _requests.push_back(std::move(r));
+        _responsens.push_back(std::move(p));
+        _enequeued.signal();
+        return f;
+    });
+}
+
+ss::future<> append_entries_buffer::stop() {
+    _enequeued.broken();
+    return _gate.close();
+}
+
+void append_entries_buffer::start() {
+    (void)ss::with_gate(_gate, [this]() mutable {
+        return ss::do_until(
+          [this] { return _gate.is_closed(); },
+          [this]() mutable {
+              return _enequeued.wait([this] { return !_requests.empty(); })
+                .then([this]() mutable { return flush(); })
+                .handle_exception_type(
+                  [](const ss::broken_condition_variable&) {
+                      // ignore exception, we are about to stop
+                  });
+          });
+    });
+}
+
+ss::future<> append_entries_buffer::flush() {
+    auto requests = std::exchange(_requests, {});
+    auto response_promises = std::exchange(_responsens, {});
+
+    return _consensus._op_lock.with(
+      [this,
+       requests = std::move(requests),
+       response_promises = std::move(response_promises)]() mutable {
+          return do_flush(std::move(requests), std::move(response_promises));
+      });
+}
+
+ss::future<> append_entries_buffer::do_flush(
+  request_t requests, response_t response_promises) {
+    bool needs_flush = false;
+    std::vector<reply_t> replies;
+    replies.reserve(requests.size());
+    for (auto& req : requests) {
+        if (req.flush) {
+            needs_flush = true;
+        }
+        try {
+            auto reply = co_await _consensus.do_append_entries(std::move(req));
```

Comment:
i don't get this. don't you need `if !req.flush; append` ? otherwise i don't see any batching here.